### PR TITLE
Feature/lab5

### PR DESCRIPTION
**Pull request by DOMZE**
Fixed settings as per @arribajuan description

### CODE REVIEW

File: Lab5/GABDemo/Startup.cs [hunk 1]
```diff
@@ -31,9 +31,17 @@ public void ConfigureServices(IServiceCollection services)
...
+            services.Configure<KeysOptions>(options =>
             {
+                options.ComputerVision = new ComputerVisionOptions
+                {
+                    ApiEndPoint = Configuration["Keys:ComputerVision:ApiEndPoint"],
+                    ApiKey = Configuration["Keys:ComputerVision:ApiKey"]
+                };
+                options.Storage = new StorageAccountOptions { ConnectionString = Configuration.GetConnectionString("ApplicationStorage") };
+            })
+            .PostConfigure<KeysOptions>(options =>
```
Comment (Carl-Hugo) [inline Lab5/GABDemo/Startup.cs]: Why call `PostConfigure` right after `Configure`? Is there a reason why not to put the two guards clause into the `Configure` method directly?

Moreover, why not create and set those guards in the `ComputerVisionOptions` constructor?
It would be more robust if `ComputerVisionOptions` guarded itself against `null` instead of programming that logic into the composition root (which could lead to that logic being overridden in a test project for example).

File: Lab5/GABDemo/Startup.cs [hunk 2]
```diff
@@ -31,9 +31,17 @@ public void ConfigureServices(IServiceCollection services)
...
-            services.Configure<KeysOptions>(Configuration.GetSection("Keys"))
-                    .PostConfigure<KeysOptions>(options =>
+            services.Configure<KeysOptions>(options =>
             {
+                options.ComputerVision = new ComputerVisionOptions
+                {
+                    ApiEndPoint = Configuration["Keys:ComputerVision:ApiEndPoint"],
+                    ApiKey = Configuration["Keys:ComputerVision:ApiKey"]
+                };
+                options.Storage = new StorageAccountOptions { ConnectionString = Configuration.GetConnectionString("ApplicationStorage") };
```
Comment (Carl-Hugo) [inline Lab5/GABDemo/Startup.cs]: What does *Keys* mean in `KeysOptions`? It is not clear.

If both `ComputerVisionOptions` and `StorageAccountOptions` are related to `KeysOptions`, why extract the `ApplicationStorage` connection string out of the `Keys` object (in *appsettings.json*)?

If the connection string does not belong to `Keys` (in *appsettings.json*), it should not belong to `KeysOptions` either, which should lead to the deletion of that object and to load both `ComputerVisionOptions` and `StorageAccountOptions` individually (both options are unrelated).

**Review by Carl-Hugo — COMMENTED** [review]
Those are notes from the software engineer in me; lets see what the others have to say :wink:

---

**Note about controller and services of lab5 (not present in this PR):**
You should considered injecting `StorageAccountOptions` in `BlobStorageManager` and injecting `ComputerVisionOptions` in `ImageAnalyzer` instead of the strings.

> You could also inject `IOptions<StorageAccountOptions>` if you prefer, like you did in `AnalyzerController`, but it does not add value; `IOptions<TOptions>` is pretty much there only for backward compatibility. [more info on options](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-2.2)

By doing that, I believe that it would be a more robust design. Moreover, it would allow you to inject the services into the `AnalyzerController` instead of `new`-ing those two objects; this would help remove the [Control Freak anti-pattern](https://subscription.packtpub.com/book/application_development/9781787121300/9/ch09lvl1sec57/anti-patterns) from the code.

To push a bit more in that same way, adding an `IImageAnalyzer` and `IBlobStorageManager` interface would also help decouple `AnalyzerController` from the services implementation.

Note: I do understand that this is only a demo; but if it takes only 5 minutes to make it better, why not!

My two cents

**Comment by DOMZE** [comment]
Fair points...

I was trying not to focus too much on the "architecture and layout" and get to the business, in fact, i could have put all the code directly in the controller but for sake of clarity is went a little further. Granted, I could have gone all the way.

For the `PostConfigure` was trying to use a bit of the ASP.NET Core 2.2 features so that ones who haven't used it can have a feeling at the same time.

For the options, I made the split. Initially, the `KeysOptions` was the structure of the initial design, so that the `Configure<>` could read directly from the configuration using the same settings structure.

I made the changes, just need adjust the README to go along with it. Will finish this tonight

**Comment by Carl-Hugo** [comment]
> I was trying not to focus too much on the "architecture and layout" and get to the business, in fact, i could have put all the code directly in the controller but for sake of clarity is went a little further. Granted, I could have gone all the way.

True and thanks, my eyes would have bleed otherwise 😉! 

> For the options, I made the split. Initially, the `KeysOptions` was the structure of the initial design, so that the `Configure<>` could read directly from the configuration using the same settings structure.

What I meant here is:

1. The name `KeysOptions` is not self-explanatory. To me, it sounds like an object containing multiple keys, while in fact, it contains, by association, one key, one endpoint, and one connection string. This could be confusing for other people.
1. The split does make sense indeed, but why keep the `KeysOptions` aggregate around when the configuration object has changed? This is a detail, I concur but might be worth taking 2 minutes to update.

Conclusion: find a better name or remove `KeysOptions`?

**Comment by mbernard** [comment]
Personally, I think we should inject the configuration object in the services. It will reduce the code and make it cleaner, plus it's best practices anyways. Doing so will also simplify the modifications need afterward in lab7 where we use a new configuration provider to get the secrets from key vault

**Comment by Carl-Hugo** [comment]
> Personally, I think we should inject the configuration object in the services. It will reduce the code and make it cleaner, plus it's best practices anyways. Doing so will also simplify the modifications need afterward in lab7 where we use a new configuration provider to get the secrets from key vault

We could even inject the dependencies directly into the services by doing something like:

```csharp
public class BlobStorageManager
{
    private readonly CloudStorageAccount _storageAccount;
    public BlobStorageManager(CloudStorageAccount storageAccount)
    {
        _storageAccount = storageAccount ?? throw new ArgumentNullException(nameof(storageAccount));
    }
    // ...
}
```

And 

```csharp
public class ImageAnalyzer
{
    private readonly ComputerVisionClient _computerVision;
    public ImageAnalyzer(ComputerVisionClient computerVision)
    {
        _computerVision = computerVision ?? throw new ArgumentNullException(nameof(computerVision));
    }
    // ...
}
```

Strictly speaking, I don't see why the responsibility of creating the `ComputerVisionClient` or the `CloudStorageAccount` should be delegated to their users; the services should only consume them. This way both services would become open for extension, but closed for modification leaving the object trees to be assembled in the *Composition Root*, as it should be when using DI; leaving it easier to evolve the system (which includes lab7).

**Comment by DOMZE** [comment]
I've commited the changes to satisfy all y'all.

Options are directly passed into the classes
Null and Empty checks are done in those as well
DI is now used to inject those classes into the controller
The options are now split into connectionStrings and regular options

**Comment by Carl-Hugo** [comment]
First thing first, nice update, we are getting there!

If the code is ok for you, it is a demo, so I'd be happy with it, but...

I don't know if we meet (sry, bad with names and to recognize toon-ish faces :wink:) and I'd prefer you not to hate me before we do (if we haven't), but if you can take some more feedback on your code, please continue to read.

---

So if you read this, I guess that you don't mind going a bit further. So here we go with a few details to make that code closer to top notch:

1. In `AnalyzerController.Index()`, you use `_` as the `Select`'s predicate variable's name; nowadays `_` is used as a kind of "null/void/absent" argument, which could be confusing. I'd change that for anything else ranging from `x` to `file`.
2. In the `BlobStorageManager`'s constructor, you use two styles of code: with and without accolades; I'd say that it would be better to pick a style and stick with it. Personally, I am old school and put accolades everywhere, but do as you wish!
3. In the `ImageAnalyzer.AnalyzeAsync` method, you explicitly typed the `analysis` variable instead of using `var`; everywhere else you seem to prefer `var`, so once again, I'd say use only one style and change `ImageAnalysis` for `var` (which is the recommended style).

It may sound picky, but I guess that those details jump into my face and I can not, not see them, sry; I have not even cloned the project.

---

Finally, **the thing that bothers me the most** would be the complexity of the `Startup.ConfigureServices`. I don't think that it should require 30 lines of code to register ±5 dependencies, I believe that you rely too much on "controlling the creation of your objects". Instead of relying on factories (`Func<IServiceProvider, TService>`) and `new`-ing your dependencies, you should register your dependencies and sub-dependencies with the IoC container and leave it do the composition "magic" for you.

> **Analogy:** DI should be a stress-less process, you give a few directions to your employee (the container), and you leave it to work for you: you delegate. Controlling dependencies is not good; it stresses the application, making it harder to maintain and to evolve :wink: 

Moreover, is there a reason to register your dependencies with a `Scoped` lifetime instead of a `Singleton` lifetime? Both services don't have any internal state, so `Singleton` would be more optimal, wouldn't it?

---

And voilà! I guess I'm done now.

*If you find this wall of text hard to follow, I can mark the code in the PR, just ask and it will be my pleasure.*

**Comment by DOMZE** [comment]
@Carl-Hugo feel free to create a PR. This is a simple demo and not production million dollar code. The idea again is to show the users how to consume the services. It should technically be simple for one to follow without having too many patterns involved because you lose the audience into what you are trying to do. I remember that some of the people doing the bootcamp last year were not 100% familiar with everything and it's a big day so overwhelming them with code that will be forgotten in a month from now is not the idea here.

**Comment by Carl-Hugo** [comment]
> not production million dollar code

It is not, indeed.

> feel free to create a PR

Sure thing, it will be way faster to code it than it was to write about it. 
My objective was to give you my feedback so that you could learn from them or correct mistakes that you missed (when you are into something, it is easy to miss some details).

Case closed!
